CompoundEditor : Save an editor's node set driver in layouts

themissingcow opened the pull request:
Allows the link state of editors to be saved within a layout and updates the default layouts such that:

 - The `Hierarchy Viewer` is linked to the `Viewer`.
 - The `Animation Editor`, `Scene Inspector` and `Primitive Inspector` are linked to the `Hierarchy Viewer`.

### User facing changes :

 - Layouts will now store and recall the link/follow state of editors.
 - Default layouts now load with several common editors linked to the Viewer.
 - Viewers can now follow other Viewers.
 - When using the pinning menu to link editors, the target editor will be visually highlighted on screen when its menu item is hovered over.
 - When hovering on the editor link swatch for a master editor, its slaves will be highlighted.
 - When hovering on the editor link swatch for a slave editor, its master will be highlighted.
 - Editor link colors will now be stable across script loads and layout changes.
 - Older versions of Gaffer are incompatible with layouts saved from this version unless `contrib/startup/gui/layoutCompatibility.py` is installed on `$GAFFER_STARTUP_PATHS`.
### API changes :

  - `GafferUI.CompoundEditor` constructor kwargs have been made private. These should only ever be used by the layout system.

  - `GafferUI.Menu` adds support for the `enter` and `leave` fields (supplied  through an IECore.MenuItemDefinition) to provide callback functionality as the user hovers over menu items presented on screen.

 - Adds `GafferUI._qtObjectIsValid` which when given an python wrapper will return if the underlying c++ object is still valid.

themissingcow commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -53,27 +53,28 @@
 class CompoundEditor( GafferUI.Editor ) :
 
-	def __init__( self, scriptNode, children=None, detachedPanels=None, windowState=None, **kw ) :
+	# The CompoundEditor constructor args are considered 'private', used only
```
Just (attempting to )recall the conversation John and I had before we left. Iirc, its that in the long-term we'd like to move to using `GraphComponents` to store layout, etc... such that we have a mechanism to store any state editors may have. It would also mean we could move to `exec` over `eval`, and so have more robust failure modes. At this point, it'd just be like a normal Gaffer script, and so were piece-meal assemble the layouts as you mentioned. So I think yeah, the inference was that it'd be better to re-build layouts as we do scripts.

johnhaddon replied:
The `children` argument has always been kindof defacto private, in that it must be passed a data structure that is nontrivial, undocumented, and has changed over time. The only way I've been aware of anyone constructing a CompoundEditor is by building the one they want manually and then saving it as a layout; there are no public APIs to allow a layout to be specified programmatically, and I've encouraged Tom not to add one for now. Ditching the `children` argument in favour of the even more opaque `_state` argument is just a way of making this decision more explicit.

Longer term we have a bigger question about how we should really be serialising these layouts, particularly if they end up serialised in scripts and folks start wanting even more state to be included. That's where I suspect a formal setup involving GraphComponents and our standard serialisation mechanism could be a win over the current ad hoc approach. It's also the point when I think we'd be happier to expose a public API for building layouts. But in the meantime, we're making things more explicitly opaque and private so we have less baggage to deal with in any future transition.

themissingcow commented:
>   * The _Viewer_ is still showing the Cube... wait, what? They were all linked together... why did one of them separate?

A good point indeed. It seems like "dragging something in to update what you see" would be way more commonly useful than "dragging something into update the pinning state of just that editor". 

Switched behaviour in https://github.com/GafferHQ/gaffer/pull/3360/commits/32bfedebbd8d9c69f2d14ebbc7d8b6e6e63f49e7 such that dragging to any linked editor updates the 'master', rather than un-linking the one you dropped on.

> Similarly, I can link my _SceneInspector_ to my _HierarchyView_, but I can't link my _HierarchyView_ to my _SceneInspector_. Why not? It feels commutative to me. Either they're sharing the same NodeSet or they aren't. I want to initiate the link behaviour from the pane my mouse is closest to.

I guess maybe the biggest thing to make sure that is understood is that 'Link' is not bi-directional. One is a master and the others are slaves. Maybe link isn't the right word, and we should just use 'Follow' - I've updated this over in the `bookmarkSets` PR here: https://github.com/GafferHQ/gaffer/pull/3323/commits/44fe7b8f51e7eaa7a82bfc3abf80d9304ca53b5c so we can see how that feels.

Otherwise, theres no technical reason not to, just listing every `NodeSetEditor` in those menus quickly made them feel overwhelming. There is an artificial limitation at the moment to only show `Viewers` and `HierarchyViews`. I picked those as they are the only two editors that let you interact with the context's _scene_ selection. In all conversations I'd had with people, they always talked about those two types of editor seeming more 'primary' than the others - which seemed more like 'accessory' editors. In that, you are either working with a `Viewer` or `Hierarchy Editor` to control what you see in other editors?

> > This version is incompatible with older layouts and layouts from older versions are incompatible with this version unless contrib/startup/gui/layoutCompatibility.py is installed on $GAFFER_STARTUP_PATHS. Which ensures full, bi-directional compatibility.
> 
> Is there a reason we're not installing that compatibility script by default?

We could do, just peoples adoption rates seem facility specific, if we just install it and make it work, we don't really think about where we are with it. Seemed like people adding it in makes it more obvious whats going on. It'd be great to remove all that compatibility code from the release as and when we can as it makes future development more convoluted.

andrewkaufman commented:
> Switched behaviour in 32bfede such that dragging to any linked editor updates the 'master', rather than un-linking the one you dropped on.

Thanks.

>> Similarly, I can link my SceneInspector to my HierarchyView, but I can't link my HierarchyView to my SceneInspector. Why not? It feels commutative to me. Either they're sharing the same NodeSet or they aren't. I want to initiate the link behaviour from the pane my mouse is closest to.

> I guess maybe the biggest thing to make sure that is understood is that 'Link' is not bi-directional. One is a master and the others are slaves. Maybe link isn't the right word, and we should just use 'Follow' - I've updated this over in the bookmarkSets PR here: 44fe7b8 so we can see how that feels.

Yeah, if we're saying its intentionally a master/slave relationship, then Follow sounds better than Link to me. But, as I say, I don't understand why its one way really. I liked the idea of bi-directional linking, and of managing the "what is viewed/edited" part from whichever linked pane is closest to my mouse. From chatting in person with Tom, it sounds like the biggest hurdle to that approach is just deciding how to effectively communicate the link & pin state using icons. There didn't seem to be any fundamental issue with allowing the "slave" to affect the "master". I think we can wait for more user feedback before continuing the discussion though.

>> Is there a reason we're not installing that compatibility script by default?

> We could do, just peoples adoption rates seem facility specific, if we just install it and make it work, we don't really think about where we are with it.

I'd argue this approach makes things more difficult for end users, which is not desirable for adoption.

> It'd be great to remove all that compatibility code from the release as and when we can as it makes future development more convoluted.

I think its out-of-scope for this PR. We should (a) design a mechanism for starting Gaffer without loading the compatibility scripts, (b) run the unit tests both with and without the compatibility layer (to make sure Gaffer itself isn't relying on them), and (c) come up with a convention for when to drop compatibility (eg every 2 major versions), so we can time-box the maintenance burden.

themissingcow commented:
@andrewkaufman See if you think the new pre-link highlighting helps, plus hovering on the swatch.

andrewkaufman commented:
> new pre-link highlighting

I like it

> hovering on the swatch

I like it, but maybe it activates/draws a bit too quickly? If I take a path over the swatch to click the pin, my screen flashes at me quite a bit.

>>> Is there a reason we're not installing that compatibility script by default?

>> We could do, just peoples adoption rates seem facility specific

> I'd argue this approach makes things more difficult for end users

Where are we at on this debate? I still think its worth installing the config by default so things "just work" and we can spinoff another Issue regarding best practices for compatibility script deprecation.

andrewkaufman commented:
Oh, I just shutdown and realized I have a bunch of these in my terminal:

```
Traceback (most recent call last):
  File "/home/andrewk/apps/gaffer/0.54.hack.0dev/cent7.x86_64/cortex/10/gaffer/python/Gaffer/WeakMethod.py", line 67, in __call__
    return m( *args, **kwArgs )
  File "/home/andrewk/apps/gaffer/0.54.hack.0dev/cent7.x86_64/cortex/10/gaffer/python/GafferUI/Menu.py", line 666, in __actionHovered
    self.__doLeave()
  File "/home/andrewk/apps/gaffer/0.54.hack.0dev/cent7.x86_64/cortex/10/gaffer/python/GafferUI/Menu.py", line 682, in __doLeave
    name = self.__lastHoverAction.data()
RuntimeError: Internal C++ object (PySide2.QtWidgets.QWidgetAction) already deleted.
```

themissingcow commented:
> Oh, I just shutdown and realized I have a bunch of these in my terminal:
> 
> ```
> Traceback (most recent call last):
>   File "/home/andrewk/apps/gaffer/0.54.hack.0dev/cent7.x86_64/cortex/10/gaffer/python/Gaffer/WeakMethod.py", line 67, in __call__
>     return m( *args, **kwArgs )
>   File "/home/andrewk/apps/gaffer/0.54.hack.0dev/cent7.x86_64/cortex/10/gaffer/python/GafferUI/Menu.py", line 666, in __actionHovered
>     self.__doLeave()
>   File "/home/andrewk/apps/gaffer/0.54.hack.0dev/cent7.x86_64/cortex/10/gaffer/python/GafferUI/Menu.py", line 682, in __doLeave
>     name = self.__lastHoverAction.data()
> RuntimeError: Internal C++ object (PySide2.QtWidgets.QWidgetAction) already deleted.
> ```

Hopefully sorted in https://github.com/GafferHQ/gaffer/pull/3360/commits/7cf64464be94a71d25fcd2e85c4d8d2c52192df5

themissingcow commented:
Hey @andrewkaufman, thanks for checking it out so quickly...

> > new pre-link highlighting
> 
> I like it

👍 

> > hovering on the swatch
> 
> I like it, but maybe it activates/draws a bit too quickly? If I take a path over the swatch to click the pin, my screen flashes at me quite a bit.

In https://github.com/GafferHQ/gaffer/pull/3360/commits/516a261f86c55ba9ee2dd25b6e9f07e17ef30963 I've moved it into the tooltip request, which means it's timed with that. Hopefully that solves the flashes without complicating the code too much. @johnhaddon is this too heinous? 

> > > > Is there a reason we're not installing that compatibility script by default?
...
> Where are we at on this debate? I still think its worth installing the config by default so things "just work" and we can spinoff another Issue regarding best practices for compatibility script deprecation.

Happy to do that, @johnhaddon?

johnhaddon commented on python/GafferUI/NodeSetEditor.py:
```diff
@@ -82,15 +82,22 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
+	# set, such that calling `getNodeSetDriver` in the nodeSetChangedSignal
+	# will return the departing driver. TODO: We need to work out a sensible
+	# way to signal once state has stabilised.
+	def setNodeSet( self, nodeSet, removeDriver = True ) :
```
I'm not convinced that this is a good API change. In the past we justified the additional complexity of drivers by saying that they were a layer on top of the (original, nice, simple) nodeSet API, but here we're muddying things by mixing the two together. I'm also unconvinced that it's clearer from a readability perspective; without it I would call `editor.drivingEditor().setNodeSet( x )`, making it absolutely clear that I'm intentionally modifying the _driver's_ state. With it, I call `editor.setNodeSet( x, removeDriver = False )`, giving the impression that I'm still operating on `editor`, which I am not.

If I'm reading correctly, it seems that the following two commits can be implemented pretty easily without this one, so I suggest we just drop it.

p.s. As an aside, one thing I like about the "Editors are Nodes" concept we've been knocking about is that we could represent the driver/driven concept nicely using connections between plugs on the editors.

themissingcow replied:
The only snag there is that `editor.drivingEditor().setNodeSet( x )` would raise a `RuntimeError` if the editor isn't linked as `drivingEditor()` returns `None`.

So the code we duplicate wherever we do this is:

```
toUpdate = editor
if editor.drivingEditor() is not None :
   toUpdate = editor.drivingEditor()
toUpdate.setNodeSet( n )
```
Which isn't the end of the world, and there are 5 different ways you can write that more concisely. I get what you mean about it looking like your modifying 'editor' but not (though ultimately are in a way)

johnhaddon replied:
Yep, I knew about that, but I stand by comments. It's just a two liner instead of a one liner :

```
target = editor.driverEditor() or editor
target( n )
```

themissingcow replied:
FIxed in https://github.com/GafferHQ/gaffer/pull/3360/commits/35a7ee2b1f9e3ce2739abf73a9d015c346b698b6 and https://github.com/GafferHQ/gaffer/pull/3360/commits/d6f7e6feda28a9f1f3c5a76dc33caa0de8102d7d

johnhaddon commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -290,6 +290,14 @@ def __captureEditorState( self ) :
 				}
+			else :
+				nodeSet = n.getNodeSet()
+				# We don't want to couple the layout-centric serialisation that assumes
```
Is this comment trying to explain why we're not using `repr( nodeSet )`?

themissingcow replied:
Improved in https://github.com/GafferHQ/gaffer/pull/3360/commits/45bc63ed9e4b00322775a27def17db7bb41f95c8

johnhaddon commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -300,20 +308,31 @@ def __restoreEditorState( self, editorState ) :
-							error = "%s: %s" % ( type(e), e )
-						)
+
+				elif "nodeSet" in state and isinstance( editor, GafferUI.NodeSetEditor ) :
```
Why `isinstance( editor, GafferUI.NodeSetEditor )`? Isn't it impossible to serialise `nodeSet` for anything but a `NodeSetEditor`?

themissingcow replied:
It is indeed. Good post, thanks. https://github.com/GafferHQ/gaffer/pull/3360/commits/45bc63ed9e4b00322775a27def17db7bb41f95c8

johnhaddon commented on python/GafferUI/Menu.py:
```diff
@@ -636,3 +656,35 @@ def keyPressEvent( self, qEvent ) :
+
+		self.__doLeave()
+
+		self.__lastHoverAction = weakref.ref( action )
```
Why a weakref? The menu owns the action already - is there any harm in owning another reference? Weak references are generally for breaking circular references, but I can't see one here. Am I missing something?

themissingcow replied:
I was getting 'underlying c++ object already deleted' errors on exit without this.

themissingcow replied:
Totally sorted in https://github.com/GafferHQ/gaffer/pull/3360/commits/34f015d55ce9017418e74dc1208ce6077a69a2b2

johnhaddon commented on python/GafferUI/Menu.py:
```diff
@@ -368,6 +375,11 @@ def __buildAction( self, item, name, parent ) :
 			qtAction.setStatusTip( item.description )
 
+		if hasattr( item, 'enter' ) :
+			parent.enterCallbacks[ name ] = item.enter
```
Why do we need to store these separately, and why on the parent? Why can't we retrieve them from the `qtAction.__item` that we've already stashed on line 322/327?

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3360/commits/34f015d55ce9017418e74dc1208ce6077a69a2b2 sorts this all out. Long story short, randomly added attributes were being dropped from the object. Deriving our own action sorted this.

johnhaddon commented on startup/gui/editor.py:
```diff
@@ -84,7 +84,9 @@ def __addFollowMenuItem( menuDefinition, editor, targetEditor, subMenuTitle, mod
 			"active" : not editor.drivesNodeSet( targetEditor ) and not isCurrent,
-			"checkBox" : isCurrent
+			"checkBox" : isCurrent,
+			"enter" : lambda : weakTarget().parent().setHighlighted( True ),
```
I like!

johnhaddon commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -1661,8 +1663,35 @@ def getToolTip( self ) :
 			for d in drivenEditors :
 				toolTipElements.append( " - _%s_" % d.getTitle() )
 
+		self.__highlightRelatedEditors( editor, True )
```
Outrageous! I suppose it's highly unlikely that anyone will trigger chaos by calling `getToolTip()` directly though, since this is a private widget squirrelled way down in the hierarchy. I think it warrants some sort of Hail Mary comment though...

themissingcow replied:
Added disclaimer in https://github.com/GafferHQ/gaffer/pull/3360/commits/08163114dcd10e1c195d06d7e93e25062af0e8a0

johnhaddon commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -1661,8 +1663,35 @@ def getToolTip( self ) :
+			toHighlight.extend( editor.drivenNodeSets( recurse = True ).keys() )
+
+		for e in toHighlight :
+			if e._qtWidget().isHidden() :
```
I definitely think this highlighting behaviour is a great feature, but I'm finding the specifics of the logic here a bit confusing in practice. When the editor is the current tab, I'm seeing the tab outline highlighted as well as the container around it. And when the editor isn't the current tab, I'm seeing just the container highlighted, which is a little non-specific. How feasible is it to highlight the tab _labels_ for all active tabs, in addition to the container outline? That way you could see when multiple editors in a container are affected, and you can see _which_ editor is affected when it's not the current one.

johnhaddon replied:
Alternatively, if we could highlight the tab outlines for all affected tabs, and leave the container alone, that would have a similar effect...

themissingcow replied:
> When the editor is the current tab, I'm seeing the tab outline highlighted as well as the container around it

It's actually a little different. You only see the dual outline when there is both the current tab, and a non-current tab both being highlighted (which they are in the default layout). If its just the current tab on its own, then you only see that one highlighted.   The alternative was to not highlight non-visible tabs, which is cleaner, but its confusing if it says its linked, then nothing gets highlighted as they're non-current.

> How feasible is it to highlight the tab labels for all active tabs, in addition to the container outline? 

Sadly with Qt its a real pain, the only pseudo-classes you have in CSS are current tab and 'the rest'. There is, to the best of my knowledge from my various endeavours in Qt of late, zero way to target individual, specific tabs.

themissingcow replied:
I was contemplating a fainter highlight for the outer container for non-visible tabs, which would look better, but then it would mean splitting the use of `setHighlighted` and manual style hackery, or introducing a new public highlight state that was pretty, (or adding `secondary=True` or something to `setHighlighted`.

johnhaddon replied:
> Sadly with Qt its a real pain, the only pseudo-classes you have in CSS are current tab and 'the rest'. There is, to the best of my knowledge from my various endeavours in Qt of late, zero way to target individual, specific tabs.

I suspected that might be the case. I was wondering if we could monkey with the tab label in some way to highlight the text, but maybe it's not worth the bother right now.

> I was contemplating a fainter highlight for the outer container for non-visible tabs, which would look better, but then it would mean splitting the use of setHighlighted and manual style hackery, or introducing a new public highlight state that was pretty, (or adding secondary=True or something to setHighlighted.

If we wanted to add API for this, I think `TabbedContainer.setHighlighted( child, highlighted )` would be the way to go, to provide a way of requesting highlight of individual tabs, whether current or not.

Let's leave this till last - we can mull it over while taking care of the rest.

johnhaddon commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -1655,7 +1655,7 @@ def getToolTip( self ) :
 
 		if masterEditor is not None :
-			toolTipElements.append( "Following _%s_" % masterEditor.getTitle() )
+			toolTipElements.append( "Following _%s_\n" % masterEditor.getTitle() )
```
Unrelated to this change, but related to the tooltip generally...

 In playing with the linking, I've found it confusing that the swatch and the pinning icon are different widgets, with different tooltips and different clickability. Given that they're both providing a representation of the same thing, it feels like they should share all the same interaction patterns, and you should be able to click the swatch to break the link and popup the menu. The linking is quite fiddly anyway, and requiring folks to hover the swatch for status but use the separate button for editing makes it a fair bit harder to interact with.

themissingcow replied:
Agreed. I don't really like the whole thing, but conscious of rabbit holes given the other priorities of late. How much time do we want to spend on this?

johnhaddon replied:
Maybe we should poll some users? I'm pretty sure though, that if we don't do it now, we won't do it for a long time...

themissingcow commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -53,27 +53,28 @@
-	def __init__( self, scriptNode, children=None, detachedPanels=None, windowState=None, **kw ) :
+	# The CompoundEditor constructor args are considered 'private', used only
+	# by the persistent layout system.
+	def __init__( self, scriptNode, _state={} ) :
```
I tried `__state` initially, but I think it gets name mangled and so makes it somewhat unintuitive to get at later.

johnhaddon replied:
Cool, let's not worry then.

themissingcow commented on contrib/startup/GafferUI/layoutCompatibility.py:
```diff
@@ -40,23 +40,45 @@
-	def __initWrapper( originalInit ) :
+		# As we only understand specific kwargs, delete any we don't know about
+		# We never used **kwarg anywhere to pass args down so this should be safe
+		kwargs = { k:v for k, v in kwargs.items() if k in initSpec.args }
```
I think this *might* make compatibility horrendous if we have to support `**kw`. As we need this compatibility shim to prevent local _and_ parent classes from barfing.  By not supporting it, we can simply introspect the locally supported args, we then don't need the version/arg matrix got all possible versions to know what should be there and the shim works across everything. There are already 3 permutations from 53 to this. Given we said we wanted to consider 'anything other than a default construct (with scriptNode) of the CompondEditor to be considered 'private', how important is honouring the widget convention for `**kw`?

themissingcow replied:
Im talking rubbish - we can still introspect things with `**kw`, so all is well.

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3360/commits/f5dfb0f194c75f5732739bbc97e2339d566f4877 reinstates this.

themissingcow commented on startup/gui/layouts.py:
```diff
@@ -65,10 +65,10 @@
 
-layouts.add( "Standard", "GafferUI.CompoundEditor( scriptNode, children = ( GafferUI.SplitContainer.Orientation.Vertical, 0.953556, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.699552, ( ( GafferUI.SplitContainer.Orientation.Vertical, 0.479263, ( {'tabs': (GafferUI.Viewer( scriptNode ),GafferSceneUI.UVInspector( scriptNode ) ), 'tabsVisible': True, 'currentTab': 0, 'pinned': [False, Fals ...
-layouts.add( "Scene", "GafferUI.CompoundEditor( scriptNode, children = ( GafferUI.SplitContainer.Orientation.Horizontal, 0.772595, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.256020, ( {'tabs': (GafferSceneUI.HierarchyView( scriptNode ),), 'tabsVisible': True, 'currentTab': 0, 'pinned': [False]}, ( GafferUI.SplitContainer.Orientation.Vertical, 0.500726, ( ( GafferUI.SplitContainer.Orien ...
+layouts.add( 'Standard', "GafferUI.CompoundEditor( scriptNode, _state={ 'children' : ( GafferUI.SplitContainer.Orientation.Vertical, 0.953488, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.699857, ( ( GafferUI.SplitContainer.Orientation.Vertical, 0.479580, ( {'tabs': (GafferUI.Viewer( scriptNode ), GafferSceneUI.UVInspector( scriptNode )), 'tabsVisible': True, 'currentTab': 0, 'pinned': [ ...
```
> This seems to be set up so that the HierarchyView follows the Viewer, and then the AnimationEditor, SceneInspector and PrimitiveInspector follow the HierarchyView. Questions :
> 
> * Why not just link everything directly to the viewer? It took me a while to figure out what was happening, and why unlinking the hierarchy view didn't do what I expected.

A common practice in production is to 'pin off' the viewer, to avoid display overhead. If everything was pinned to the viewer, then those editors all now need re-pining to be usable.

The idea was that when the viewer is pinned off, the Hierarchy View is now the authoritative source of scene selection, etc...  so following that is the next logical thing. With this setup, if you just unlink the hierarchy view from the viewer when its pinned off, there is minimal house keeping.

> * Should the AnimationEditor really follow the Viewer? As an _editor_, I would expect it to follow the node selection, just as the NodeEditor does.

Took a punt on that one, can easily unlink :)

johnhaddon replied:
> A common practice in production is to 'pin off' the viewer, to avoid display overhead.

I think this dates from the days before the pause button, which I think is a more direct of expressing the user's intention there. I'd like to encourage folks to use the pause button instead, and if that leads to some requests to improve the pause behaviour, then so be it (the obvious one being making it more obvious when we're paused). Anyone want to argue the opposite?

> With this setup, if you just unlink the hierarchy view from the viewer when its pinned off, there is minimal house keeping.

Yeah, I get it. I just think it's more complex than necessary for newcomers, particularly those who may not have the "pin off" habit.

> Took a punt on that one, can easily unlink :)

Cool - I think that would be best.

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3360/commits/01683b6f8ab19f25f27f440501135dca3b8fa846 Removes animation link, and links all to the viewer. This may highlight that we need to make the pause button (as that should remove the need to pin-off the viewer), but its the right direction to be going in.

johnhaddon replied:
I've made https://github.com/GafferHQ/gaffer/pull/3390 to make paused viewers more obvious, which hopefully goes some way to making this the right decision.

The Multi-Monitor layout doesn't have the SceneInspector linked to the Viewer though - can we fix that?

themissingcow commented on contrib/startup/GafferUI/layoutCompatibility.py:
```diff
@@ -40,23 +40,45 @@
+
+	def init( self, *args, **kwargs ) :
+
+		# If this version uses _state and incoming args are pre-state, nest them
```
@andrewkaufman you can rest easy now

johnhaddon replied:
In an offline discussion we've agreed to split things so that the version that installs with Gaffer itself will deal with backwards compatibility (loading old layouts), and the contrib file will remain for forwards compatibility (because it needs to be installed separately for old gaffer versions).

themissingcow commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -1630,8 +1630,7 @@ class _DrivenEditorSwatch( _Frame ) :
 		imath.Color3f( 0.85, 0.80, 0.48 ),
-		imath.Color3f( 0.62, 0.79, 0.93 ),
-		imath.Color3f( 0.27, 0.45, 0.21 ),
+		imath.Color3f( 0.35, 0.55, 0.28 ),
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3360/commits/a431e3985fd7d5a8ea9e3a3d964bcacedaf1bd1c, its now per-layout, which is the right way.

themissingcow commented:
Thanks John,

> * Fix the linking for the multi-monitor layout

Fixed in https://github.com/GafferHQ/gaffer/pull/3360/commits/f143e1133f1b9cacaae4d353b42cb18a9cef8d9e, thanks for catching that.

> * Fix the comment restricting people's sovereign right to call `getToolTip()` :)

Re-worded in https://github.com/GafferHQ/gaffer/pull/3360/commits/50dd8870c60e829985332d7c27ea49d8882cd225.

> * Put a bit of thought into how doable it is to give the link button and the swatch the same behaviour. I think it might not be all that bad.

As per our off-line discussion, will tackle this in a new PR.

> * Fix this bug that I've just found :
>   
>   1. Open Gaffer.
>   2. Switch to Scene layout.
>   3. Switch back to Standard layout.
>   4. See exception traceback in shell.

#3179 strikes again! 

Fixed via https://github.com/GafferHQ/gaffer/pull/3360/commits/012a5be60a0c09153f6427fa96f5db53ab8e6b28 in https://github.com/GafferHQ/gaffer/pull/3360/commits/e6cd51429a266409266cf0bdf4d31535d29d7c24.

ta lots.


> I've gone through and resolved all the conversations that have been addressed so far. Unfortunately I added a new comment while doing that, but GitHub wouldn't let me start a new review until the old one was closed, or something, I didn't quite understand. Here is a list of the things I think are still outstanding :
> 
> * Fix the linking for the multi-monitor layout
> * Fix the comment restricting people's sovereign right to call `getToolTip()` :)
> * Put a bit of thought into how doable it is to give the link button and the swatch the same behaviour. I think it might not be all that bad.
> * Fix this bug that I've just found :
>   
>   1. Open Gaffer.
>   2. Switch to Scene layout.
>   3. Switch back to Standard layout.
>   4. See exception traceback in shell.

johnhaddon commented:
Thanks Tom! I've merged, but I did notice further thing : the UVInspector isn't following the Viewer in any of the standard layouts. I think our rule is that "inspectors follow the viewer, editors follow the selection"? If that's right, could you make a quick PR to get it fixed?

themissingcow commented:
> Thanks Tom! I've merged, but I did notice further thing : the UVInspector isn't following the Viewer in any of the standard layouts. I think our rule is that "inspectors follow the viewer, editors follow the selection"? If that's right, could you make a quick PR to get it fixed?

@johnhaddon #3398
